File: clj-advoc/src/clj_advoc/day_7.clj
(ns clj-advoc.day-7
  (:require [clojure.java.io :as io]
            [clojure.string :as s]))


(defn problem-input
  [filename]
  (let [rdr (io/reader (io/resource filename))]
    (doall (line-seq rdr))))


(defn parse-contents
  "Take a string of bag contents and return a list of rules."
  [contents]
  (if (re-find #"no other" contents)
    []
    (as-> contents c
      (s/split c #",")
      (map #(s/replace % #"bag(s)?" "") c)
      (map s/trim c)
      (map (fn [x]
             (let [[n desc color] (take 3 (s/split x #" "))]
               [(Integer/parseInt n) (str desc " " color)]))
           c))))


(defn line->rule
  "Convert the given sentence into a rule map."
  [line]
  (let [[bag contents] (map s/trim (s/split line #"contain"))
        bag-name (-> bag
                     (s/replace "bags" "")
                     s/trim)
        bag-contents (parse-contents contents)]
    {bag-name bag-contents}))


(defn construct-rules
  [lines]
  (reduce (fn [rules line]
            (merge rules (line->rule line)))
          {}
          lines))


;; ────────────────────────────── PART 1 ──────────────────────────────

;; Use this to memoize the can-contain-gold? function
(def gold (atom {}))

(defn can-contain-gold?
  "Check whether according to given rules, whether a certain bag
  can contain a gold bag or not."
  [rules-map bag-name]
  (let [contents (get rules-map bag-name)]
    (if-not (seq contents)
      false
      (reduce (fn [ans [_ bag]]
                (or ans
                    (= bag "shiny gold")
                    (if-let [memoize (get @gold bag)]
                      memoize
                      (get (swap! gold
                                  assoc
                                  bag
                                  (can-contain-gold? rules-map bag))
                           bag))))
              false
              contents))))


(defn solution-part-1
  [inpt]
  (let [rules-map (construct-rules inpt)]
    (count (filter (partial can-contain-gold? rules-map)
                   (keys rules-map)))))

;; ────────────────────────────── PART 2 ──────────────────────────────

;; Use this to memoize the bags-count function
(def bags-count (atom {}))

(defn count-bags
  "Given a bag-name, find the total number of bags the must
  be contained in it."
  [rules-map bag-name]
  (let [contents (get rules-map bag-name)]
    (if-not (seq contents)
      0
      (reduce (fn [cnt [n bag]]
                (+ cnt
                   n
                   (* n (if-let [memoize (get @bags-count bag)]
                          memoize
                          (get (swap! bags-count
                                      assoc
                                      bag
                                      (count-bags rules-map bag))
                               bag)))))
              0
              contents))))


(defn solution-part-2
  [inpt]
  (let [rules-map (construct-rules inpt)]
    (count-bags rules-map "shiny gold")))


(defn main
  []
  (let [inpt (problem-input "day-7-part-1.input")]
    (-> inpt
        (solution-part-1)
        (println " : Answer Part 1"))
    (-> inpt
        (solution-part-2)
        (println " : Answer Part 2"))))

File: py-advoc/day_7.py
import os
import re
from functools import lru_cache

INPUT_PATH = "../inputs/"


def problem_input(filename):
    path = os.path.join(INPUT_PATH, filename)
    inpt = []
    with open(path, 'r') as f:
        for line in f:
            inpt.append(line)
    return inpt


def parse_contents(contents):
    if re.search('no other', contents):
        return []

    cc = [re.sub('bag(s)?', '', i).strip() for i in contents.split(',')]
    rules = []
    for i in cc:
        n, desc, color = i.split()[:3]
        rules.append([int(n), desc + " " + color])
    return rules



def conv_line_rule(line):
    bag, contents = [i.strip() for i in line.split('contain')]
    bag_name = bag.replace('bags', '').strip()
    bag_contents = parse_contents(contents)
    return {bag_name: bag_contents}


def construct_rules(lines):
    rules_map = {}
    for line in lines:
        rule = conv_line_rule(line)
        rules_map = {**rules_map, **rule}
    return rules_map

#  ───────────────────────────── PART 1 ─────────────────────────────
cache_rule_map = {}

@lru_cache(maxsize=1000)
def cached_can_contain_gold(bag_name):
    contents = cache_rule_map[bag_name]
    if len(contents) == 0:
        return False

    ans = False
    for n, bag in contents:
        ans = ans or bag == 'shiny gold' or cached_can_contain_gold(bag)
    return ans

def solution_part_1_cached(inpt):
    global cache_rule_map
    cache_rule_map = construct_rules(inpt)
    count = 0
    for bag_name in cache_rule_map.keys():
        if cached_can_contain_gold(bag_name):
            count += 1
    return count

# Memoized solution timing stats
# 10.4 ms ± 80.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

def can_contain_gold(rules_map, bag_name):
    """
    Check whether according to given rules, whether a certain bag
    can contain a gold bag or not.
    """
    contents = rules_map[bag_name]
    if len(contents) == 0:
        return False

    ans = False
    for n, bag in contents:
        ans = ans or bag == 'shiny gold' or can_contain_gold(rules_map, bag)
    return ans

def solution_part_1(inpt):
    rules_map = construct_rules(inpt)
    count = 0
    for bag_name in rules_map.keys():
        if can_contain_gold(rules_map, bag_name):
            count += 1
    return count

# Non-memoized solution timing stats
# 70.1 ms ± 1.29 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

#  ───────────────────────────── PART 2 ─────────────────────────────

cache_rule_map = {}
@lru_cache(maxsize=1000)
def count_bags(bag_name):
    contents = cache_rule_map[bag_name]
    if len(contents) == 0:
        return 0

    count = 0
    for n, bag in contents:
        count = count + n + (n * count_bags(bag))
    return count

def solution_part_2(inpt):
    global cache_rule_map
    cache_rule_map = construct_rules(inpt)
    return count_bags('shiny gold')
